### PR TITLE
store/tikv: collect exited goroutines when shutdown the server

### PR DESCRIPTION
**Pull request by hicqu**
### What problem does this PR solve? 
Some internal schrodinger tests about groutine leak could fail without the patch, because when the tidb server is shutdown, the main goroutine won't wait all goroutines in `connArray` exit.

### What is changed and how it works?
This PR add a wait group in `connArray`, so now we can collect all exited goroutines in `connArray`.

### Check List 

Tests 
A schrodinger test case in #10375 .

### CODE REVIEW

File: store/tikv/client.go [hunk 1]
```diff
@@ -125,7 +130,9 @@ func (c *batchCommandsClient) batchRecvLoop(cfg config.TiKVClient) {
 				zap.Stack("stack"))
 			logutil.Logger(context.Background()).Info("restart batchRecvLoop")
 			go c.batchRecvLoop(cfg)
+			return
 		}
+		c.waitGroup.Done()
```
Comment (zz-jason) [inline store/tikv/client.go]: it should be put before the `if` statement.

Comment (hicqu) [reply store/tikv/client.go]: No. Only if the goroutine exits without error we need to decrease the wait group.

Comment (zz-jason) [reply store/tikv/client.go]: If the goroutine exits with error, the counter in wait group won't decrease. Will the outer side wait for an infinite time under this condition?

**Comment by tiancaiamao** [comment]
When the TiDB server is shutdown, the process should exit, why should the schrodinger platform detect a goroutine leak? @hicqu

**Comment by zz-jason** [comment]
@hicqu friendly ping, any update?

**Comment by zz-jason** [comment]
closed due to outdated. fell free to reopen it again when it's needed.

**Pull request closed without merging**